transport/http/jsonrpc/client.go

```diff
@@ -0,0 +1,221 @@
+	Generate() *RequestID
+}
+
+// ClientRequestIDGenerator is executed before each request to generate an ID
+// for the request.
+// By default, AutoIncrementRequestID is used.
+func ClientRequestIDGenerator(g requestIDGenerator) ClientOption {
+	return func(c *Client) { c.requestID = g }
+}
+
+// BufferedStream sets whether the Response.Body is left open, allowing it
+// to be read from later. Useful for transporting a file as a buffered stream.
+func BufferedStream(buffered bool) ClientOption {
+	return func(c *Client) { c.bufferedStream = buffered }
+}
+
+// Endpoint returns a usable endpoint that invokes the remote endpoint.
+func (c Client) Endpoint() endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		var (
+			resp *http.Response
+			err  error
+		)
+		if c.finalizer != nil {
+			defer func() {
+				if resp != nil {
+					ctx = context.WithValue(ctx, httptransport.ContextKeyResponseHeaders, resp.Header)
+					ctx = context.WithValue(ctx, httptransport.ContextKeyResponseSize, resp.ContentLength)
+				}
+				c.finalizer(ctx, err)
+			}()
+		}
+
+		var params json.RawMessage
+		if params, err = c.enc(ctx, request); err != nil {
+			return nil, err
+		}
+		rpcReq := Request{
+			JSONRPC: "",
+			Method:  c.method,
+			Params:  params,
+			ID:      c.requestID.Generate(),
+		}
+
+		req, err := http.NewRequest("POST", c.tgt.String(), nil)
+		if err != nil {
+			return nil, err
+		}
+
+		req.Header.Set("Content-Type", "application/json; charset=utf-8")
+		var b bytes.Buffer
+		req.Body = ioutil.NopCloser(&b)
+		err = json.NewEncoder(&b).Encode(rpcReq)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, f := range c.before {
+			ctx = f(ctx, req)
+		}
+
+		resp, err = c.client.Do(req.WithContext(ctx))
+		if err != nil {
+			return nil, err
+		}
+
+		if !c.bufferedStream {
+			defer resp.Body.Close()
+		}
+
+		// Decode the body into an object
+		var rpcRes Response
+		err = json.NewDecoder(resp.Body).Decode(&rpcRes)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, f := range c.after {
+			ctx = f(ctx, resp)
+		}
+
+		return c.dec(ctx, rpcRes.Result)
+	}
+}
+
+// ClientFinalizerFunc can be used to perform work at the end of a client HTTP
+// request, after the response is returned. The principal
+// intended use is for error logging. Additional response parameters are
+// provided in the context under keys with the ContextKeyResponse prefix.
+// Note: err may be nil. There maybe also no additional response parameters
+// depending on when an error occurs.
+type ClientFinalizerFunc func(ctx context.Context, err error)
+
+// AutoIncrementRequestID is a RequestIDGenerator that generates
+// auto-incrementing integer IDs.
+type AutoIncrementRequestID int
+
+// Generate satisfies RequestIDGenerator
+func (i *AutoIncrementRequestID) Generate() *RequestID {
+	id := *i
+	*i++
+	return &RequestID{
+		intValue: int(id),
+	}
+}
```

Comment:
This isn't goroutine-safe, and the type that includes it (the client) doesn't perform any synchronization. You've got a data race! You need to either put synchronization in the client, and un-export this type or explicitly mark it as not goroutine-safe; or put synchronization in this type directly.

Reply:
OK, I think that's addressed now, with an atomic.AddUint64 in Generate().